python/tskit/trees.py

```diff
@@ -8073,6 +8122,40 @@ def genetic_relatedness(
 
         return out
 
+    def genetic_relatedness_matrix(
+        self,
+        sample_sets=None,
+        *,
+        windows=None,
+        num_threads=0,
+        mode=None,
+        span_normalise=True,
+    ):
+        D = self.divergence_matrix(
+            sample_sets,
+            windows=windows,
+            num_threads=num_threads,
+            mode=mode,
+            span_normalise=span_normalise,
+        )
+
+        def _normalise(B):
+            if len(B) == 0:
+                return B
+            K = B + np.mean(B)
+            y = np.mean(B, axis=0)
+            X = y[:, np.newaxis] + y[np.newaxis, :]
+            K -= X
+            # FIXME I don't know what this factor of -2 is about
+            return K / -2
```

Comment:
Okay, let's see - this must be because relatedness between `I` and `J` is defined [in terms of shared alleles](https://tskit.dev/tskit/docs/latest/python-api.html#tskit.TreeSequence.genetic_relatedness) as:
```
 E[m(I,J) - m(I,S) - m(J,T) + m(S,T)]
```
where `m(,)` is the number of shared alleles and `S` and `T` are independently chosen sample sets.

So, naively, if `d(I,J)` is the number of *differing* alleles, and `Z` is the total number of alleles, then
```
 m(I,J) = Z - d(I,J)
```
and so
```
m(I,J) - m(I,S) - m(J,T) + m(S,T) = - d(I,J) + d(I,S) + d(J,T) - d(S,T)
```

However: apparently for relatedness we want `m(I,J)` for sample sets `I` and `J` to be the *sum* over samples `i in I` and `j in J` of `m(i,j)`, while for genetic divergence we have that `diversity(I,J)` is the *average* over `diversity(i,j)`.

Hm, now I'm a bit turned around, since in *this* implementation, we're taking the [average over samples](https://github.com/tskit-dev/tskit/pull/2823/files#diff-1604fd5688ee5df1192527915ccc2769b428aa1bf4090f4cae5b24d05d3dea37R7612) as opposed the sum over samples. This contradicts what  [the docs for `genetic_relatedness` say](https://tskit.dev/tskit/docs/latest/python-api.html#tskit.TreeSequence.genetic_relatedness) , but then again, I think would be what someone would actually expect to get from using sample sets.

Before I try to figure out what we want to do here more, do you have a reference of previous discussion on this?

Reply:
Well I just got the 2 by looking at the results and guessing. The implementation is based on [this gist](https://gist.github.com/jeromekelleher/33927b941f2d63317049aacce16ff63a) where I went back to first principles on the GRM. I can't remember where I got the normalisation expression from though I'm afraid.